src/mesh/mesh_refinement_ops.hpp

```diff
@@ -0,0 +1,245 @@
+KOKKOS_INLINE_FUNCTION Real GetX<3>(const Coordinates_t &coords, int i) {
+  return coords.x3v(i);
+}
+template <int DIM>
+KOKKOS_FORCEINLINE_FUNCTION void
+GetSlopes(const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+          const IndexRange &cib, const IndexRange &ib, int i, int &fi, Real &dxm,
+          Real &dxp, Real &dxfm, Real &dxfp) {
+  fi = (i - cib.s) * 2 + ib.s;
+  const Real xm = GetX<DIM>(coarse_coords, i - 1);
+  const Real xc = GetX<DIM>(coarse_coords, i);
+  const Real xp = GetX<DIM>(coarse_coords, i + 1);
+  dxm = xc - xm;
+  dxp = xp - xc;
+  const Real fxm = GetX<DIM>(coords, fi);
+  const Real fxp = GetX<DIM>(coords, fi + 1);
+  dxfm = xc - fxm;
+  dxfp = fxp - xc;
+}
+KOKKOS_FORCEINLINE_FUNCTION
+Real GradMinMod(const Real ccval, const Real fm, const Real fp, const Real dxm,
+                const Real dxp) {
+  const Real gxm = (ccval - fm) / dxm;
+  const Real gxp = (fp - ccval) / dxp;
+  return 0.5 * (SIGN(gxm) + SIGN(gxp)) * std::min(std::abs(gxm), std::abs(gxp));
+}
+} // namespace util
+
+template <int DIM>
+struct RestrictCellAverage {
+
+  KOKKOS_FORCEINLINE_FUNCTION static void
+  Do(const int l, const int m, const int n, const int ck, const int cj, const int ci,
+     const IndexRange &ckb, const IndexRange &cjb, const IndexRange &cib,
+     const IndexRange &kb, const IndexRange &jb, const IndexRange &ib,
+     const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+     const ParArray6D<Real> &coarse, const ParArray6D<Real> &fine) {
+    const int i = (ci - cib.s) * 2 + ib.s;
+    int j = jb.s;
+    if (DIM > 1) {
+      j = (cj - cjb.s) * 2 + jb.s;
+    }
+    int k = kb.s;
+    if (DIM > 2) {
+      k = (ck - ckb.s) * 2 + kb.s;
+    }
+    // JMM: If dimensionality is wrong, accesses are out of bounds. Only
+    // access cells if dimensionality is correct.
+    const Real vol000 = coords.Volume(k, j, i);
+    const Real vol001 = coords.Volume(k, j, i + 1);
+    const Real fine000 = fine(l, m, n, k, j, i);
+    const Real fine001 = fine(l, m, n, k, j, i + 1);
+    Real vol010, vol011, vol100, vol101, vol110, vol111;
+    Real fine010, fine011, fine100, fine101, fine110, fine111;
+    vol010 = vol011 = vol100 = vol101 = vol110 = vol111 = 0;
+    fine010 = fine011 = fine100 = fine101 = fine110 = fine111 = 0;
+    if (DIM > 1) { // TODO(c++17) make constexpr
+      vol010 = coords.Volume(k, j + 1, i);
+      vol011 = coords.Volume(k, j + 1, i + 1);
+      fine010 = fine(l, m, n, k, j + 1, i);
+      fine011 = fine(l, m, n, k, j + 1, i + 1);
+    }
+    if (DIM > 2) { // TODO(c++17) make constexpr
+      vol100 = coords.Volume(k + 1, j, i);
+      vol101 = coords.Volume(k + 1, j, i + 1);
+      vol110 = coords.Volume(k + 1, j + 1, i);
+      vol111 = coords.Volume(k + 1, j + 1, i + 1);
+      fine100 = fine(l, m, n, k + 1, j, i);
+      fine101 = fine(l, m, n, k + 1, j, i + 1);
+      fine110 = fine(l, m, n, k + 1, j + 1, i);
+      fine111 = fine(l, m, n, k + 1, j + 1, i + 1);
+    }
```

Comment:
this is technically more cycles, isn't it? Since the looping will happen regardless of `DIM`? I'd rather avoid the loops, or at least not coalesce them. But I think replacing the names with arrays is a good improvement.

Reply:
yeah, it is more cycles for 1D and 2D but I am skeptical that it makes a difference compared to memory access.

Reply:
Actually just realized this won't work because I need `vol` to be set to zero for the elements that are unused. Fortunately, I think I found a better solution that uses your loop idea:
```C++
    Real vol[2][2][2], terms[2][2][2];
    std::memset(&vol[0][0][0], 0., 8*sizeof(Real));
    std::memset(&terms[0][0][0], 0., 8*sizeof(Real));
    constexpr int xoff = 1;
    constexpr int yoff = (DIM > 1);
    constexpr int zoff = (DIM > 2);
    for (int iz = 0; iz < 1 + zoff; ++iz) {
      for (int iy = 0; iy < 1 + yoff; ++iy) {
        for (int ix = 0; ix < 1 + xoff; ++ix) {
          vol[iz][iy][ix] = coords.Volume(k + iz, j + iy, i + ix);
          terms[iz][iy][ix] = vol[iz][iy][ix]*fine(l, m, n, k + iz, j + iy, i + ix);
        }
      }
    }
```
this has the advantage of no extra cycles too, since those loops are exaclty the size we need. And better yet, the bounds are evaluated at compile time, since the offsets can be `constexpr` so we don't need `constexpr if` here to get all the advantages of the templating.

Reply:
I think the original solution will work since you are just adding the volume twice for every term you add twice (when `zoff` and/or `yoff` is zero), so the average is the same, but I agree this is probably a little faster.